include/geolib/ros/msg_conversions.h

```diff
@@ -42,6 +45,31 @@ inline void convert(const geo::Transform& t, geometry_msgs::Transform& msg) {
     convert(t.getBasis(), msg.rotation);
 }
 
+inline void convert(const TriangleI& t, shape_msgs::MeshTriangle& msg) {
+    msg.vertex_indices[0] = t.i1_;
+    msg.vertex_indices[1] = t.i2_;
+    msg.vertex_indices[2] = t.i3_;
+}
+
+void convert(const geo::Mesh& m, shape_msgs::Mesh& msg) {
+    std::vector<Vector3> points = m.getPoints();
+    std::vector<TriangleI> triangles = m.getTriangleIs();
```

Comment:
You make a full copy of the points and the triangles here. Below, you only iterate over the new vector, copying them again. Make these two lines a reference to the original data to eliminate the first copy.

That is, add an `&` like:
``` cpp
std::vector<Vector3>& points = m.getPoints();
std::vector<TriangleI>& triangles = m.getTriangleIs();
```

Reply:
This results in an error.
```
/home/amigo/ros/kinetic/system/src/geolib2/include/geolib/ros/msg_conversions.h: In function ‘void geo::convert(const geo::Mesh&, shape_msgs::Mesh&)’:
/home/amigo/ros/kinetic/system/src/geolib2/include/geolib/ros/msg_conversions.h:55:47: error: binding ‘const std::vector<geo::Vec3T<double> >’ to reference of type ‘std::vector<geo::Vec3T<double> >&’ discards qualifiers
     std::vector<Vector3>& points = m.getPoints();
                                               ^
/home/amigo/ros/kinetic/system/src/geolib2/include/geolib/ros/msg_conversions.h:56:56: error: binding ‘const std::vector<geo::TriangleI>’ to reference of type ‘std::vector<geo::TriangleI>&’ discards qualifiers
     std::vector<TriangleI>& triangles = m.getTriangleIs();
```
 ```getPoints``` and ```getTriangleIs``` already return a reference. You can't create a reference to a reference right? I am not 100% sure here. You are the expert on this one.

Reply:
Oh, missed const I see. Sorry about that.

"discards qualifiers" normally means you're changing access rights. That is at least the case here. The error message is however quite confusing if you don't know where to look. The clue is in the types. Below I wrote the 'points' types error message with aligned types for easier comparison:
```
/home/amigo/ros/kinetic/system/src/geolib2/include/geolib/ros/msg_conversions.h:55:47: error: binding
‘const std::vector<geo::Vec3T<double> >’ to reference of type
'      std::vector<geo::Vec3T<double> >&’ discards qualifiers ...
```
So what the message says is "you have read-only `const vector<..>` and you assign it to read/write `vector<..>`". If the compiler allowed this, you'd have a loop-hole where you can modify a read-only variable through the reference. Since this is unwanted, the compiler doesn't accept the assignment. There is an override for this called const cast, but that pretty much belongs in the 'avoid nuclear explosion` category.

Instead, comply with the wishes of the geolib class, and make the reference read-only too:
```
const std::vector<Vector3>& points = m.getPoints();
```
Obviously, any attempt gain write access or even write a value through the reference will now break in the code below. Luckily, this isn't the case. The points iterator is a nice `const_iterator`, and there are no other uses of `points` in this function.

A similar solution should be applied to the triangles vector.

----
The purpose of a reference is to introduce a local name for some value that is elsewhere, to avoid copying such a value. So while internally, they are much like pointers (they point to some other location in memory with the real data), at C++ level, they behave much asif the value they refer to exists locally inside the reference. Thus `int a=b` and `int &c=b` are pretty much the same, except the latter doesn't actually copy `b`.
`a = 3` and `c = 5` change `a` respectively `b` (since `c` has no storage, it's a new name for `b` only). Setting up a reference to a reference makes them both point to the real data, ie `int &d = c` makes both `c` and `d` refer to `b`.

The latter of course also happens with pointers, ie `int *p = a; int *q = p;` makes both `p` and `q` point to `a`. A pointer to a pointer has a different type, like `int **x = &p`. This stuff is totally consistent technically, except we start to get confused quite fast here.